source-s3/capture.go

```diff
@@ -0,0 +1,209 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"regexp"
+	//"strings"
+	//"time"
+
+	//"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/s3"
+	//"github.com/estuary/connectors/go-types/airbyte"
+	"github.com/estuary/connectors/go-types/parser"
+	log "github.com/sirupsen/logrus"
+)
+
+var StreamCompleted = errors.New("stream completed")
+
+type readResult struct {
+	err         error
+	bucket      string
+	streamID    string
+	relativeKey string
+	state       *objectState
+	record      json.RawMessage
+}
+
+func NewStream(config *Config, client *s3.S3, streamID string, state streamState) (*stream, error) {
+	var bucket, prefix = parseStreamName(streamID)
+	var matchKey *regexp.Regexp
+	var err error
+	if config.MatchKeys != "" {
+		matchKey, err = regexp.Compile(config.MatchKeys)
+		if err != nil {
+			return nil, fmt.Errorf("invalid regex for matchKey: %w", err)
+		}
+	}
+	return &stream{
+		streamID:  streamID,
+		bucket:    bucket,
+		prefix:    prefix,
+		state:     state,
+		client:    client,
+		config:    config,
+		matchKeys: matchKey,
+	}, nil
+}
+
+func (c *stream) Start(ctx context.Context, onResult func(readResult) error) {
+	var err = c.captureStreamInternal(ctx, onResult)
+	// Send a StreamCompleted to indicate that we've completed successfully
+	if err == nil {
+		err = StreamCompleted
+	}
+	onResult(readResult{err: err})
+}
+
+type stream struct {
+	streamID  string
+	bucket    string
+	prefix    string
+	state     streamState
+	client    *s3.S3
+	config    *Config
+	matchKeys *regexp.Regexp
+}
+
+func (c *stream) captureStreamInternal(ctx context.Context, onResult func(readResult) error) error {
+	log.Debug("Starting cature")
+	var listResults = listAllObjects(ctx, c.bucket, c.prefix, c.client)
+	var objectCount = 0
+	for result := range listResults {
+		objectCount++
+		if result.err != nil {
+			return fmt.Errorf("listing objects: %w", result.err)
+		}
+		// Should we skip importing this item because it doesn't match the configured regex?
+		if c.matchKeys != nil && !c.matchKeys.MatchString(result.object.relativeKey) {
+			continue
+		}
+		if err := c.importObject(ctx, result.object, onResult); err != nil {
+			return fmt.Errorf("failed to import object: '%s': %w", result.object.relativeKey, err)
+		}
+	}
+	log.WithField("observedObjects", objectCount).Info("completed processing stream")
+	return nil
+}
+
+func (c *stream) importObject(ctx context.Context, obj *s3Object, onResult func(readResult) error) error {
+	var fullKey = obj.fullKey()
+
+	// Does this key hash into our shard's assigned range?
```

Comment:
As I recall the plan was to hoist and consolidate these checks plus the regex check into a single place ?

Reply:
yea that was the plan, and apparently I dropped the ball there. Fixed now